library/Translate/src/Juvix/Core/Common/Context/Traverse.hs

```diff
@@ -67,24 +69,52 @@ onEntry ::
   (NameSymbol.T -> Context.Definition term ty sumRep -> t) ->
   Entry term ty sumRep ->
   t
-onEntry f (Entry {name, def}) = f name def
+onEntry f Entry {name, def} = f name def
 
 -- | Sorts a context by dependency order. Each element of the output is
 -- a mutually-recursive group, whose elements depend only on each other and
 -- elements of previous groups. The first element of each pair is its
 -- fully-qualified name.
 recGroups ::
-  (Data term, Data ty, Data sumRep) =>
+  (Show ty, Show term, Show sumRep, Data term, Data ty, Data sumRep, Eq ty, Eq term, Eq sumRep) =>
   Context.T term ty sumRep ->
   [Group term ty sumRep]
-recGroups ctx@(Context.T {topLevelMap}) =
+recGroups ctx@Context.T {topLevelMap} =
   let top = topLevelMap
       (groups, deps) = run_ ctx $ recGroups' injectTopLevel $ toNameSpace top
       get n = maybe [] toList $ HashMap.lookup n deps
       edges = map (\(n, gs) -> (gs, n, get n)) $ HashMap.toList groups
       (g, fromV', _) = Graph.graphFromEdges edges
       fromV v = let (gs, _, _) = fromV' v in gs
-   in Graph.topSort g |> reverse |> concatMap fromV
+   in Graph.topSort g |> reverse |> concatMap fromV |> sortBy orderDatatypes |> groupCons
+
+-- | Join type and data constructors in a single group
+groupCons :: (Foldable t, Eq ty, Eq term, Eq sumRep) => t (NonEmpty (Entry term ty sumRep)) -> [Group term ty sumRep]
+groupCons = fmap snd . foldr f mempty
+  where
+    f entry@(a NonEmpty.:| _as) acc
+      | Context.SumCon Context.Sum {sumTName} <- def a,
+        Just _ <- find (elem sumTName) (fst <$> acc) =
+        -- Find if type declar of a data constructor exists
+        foldr (g sumTName) mempty acc
+      | otherwise = (name a, entry) : acc
+      where
+        g sumTName (k, v) a =
+          if sumTName `elem` k
+            then (k, v `NonEmpty.union` entry) : a
+            else (k, v) : a
```

Comment:
Bit annoying how this is needed, given how you setup the sum type properly, so Constructors reference the data type already